pkg/gardenlet/controller/shoot/care/health.go

```diff
@@ -786,8 +738,67 @@ func CheckForExpiredNodeLeases(nodeList *corev1.NodeList, leaseList *coordinatio
 	return nil
 }
 
-// CheckNodesScalingUp returns an error if nodes are being scaled up.
-func CheckNodesScalingUp(machineList *machinev1alpha1.MachineList, readyNodes, desiredMachines int) error {
+// CheckNodesScaling returns an error and a string describing the scaling if there is a scaling happening
+func CheckNodesScaling(ctx context.Context, seedClient client.Client, nodeList *corev1.NodeList, machineDeploymentList *machinev1alpha1.MachineDeploymentList, controlPlaneNamespace string) (string, error) {
+	var (
+		readyAndSchedulableNodes int
+		registeredNodes          = len(nodeList.Items)
+		desiredMachines          = getDesiredMachineCount(machineDeploymentList.Items)
+	)
+
+	for _, node := range nodeList.Items {
+		if node.Spec.Unschedulable {
+			continue
+		}
+		for _, condition := range node.Status.Conditions {
+			if condition.Type == corev1.NodeReady && condition.Status == corev1.ConditionTrue {
+				readyAndSchedulableNodes++
+			}
+		}
+	}
+
+	machineList := &machinev1alpha1.MachineList{}
+	if registeredNodes == desiredMachines && readyAndSchedulableNodes == desiredMachines {
+		return "", nil
+	}
+
```

Comment:
```suggestion
	if readyAndSchedulableNodes == desiredMachines {
		return "", nil
	}
	
	machineList := &machinev1alpha1.MachineList{}
```

`registeredNodes == desiredMachines` seems redundant to me.

Reply:
I was thinking initially the same thing when reviewing.  Only counter-argument case that came to my mind is when during a rolling update we have 1 desiredMachine(for example), 2 registered nodes - 1 unready and 1 ready and schedulable. in old code we will get `too many worker nodes are registered. Exceedin ...` error and in new one we will not
1. Is this even possible - with 1 unready can the second ready node be schedulable?
2. Even if this state is transiently possible will the reporting of it even matter?

Reply:
I'm not sure about how this is related to the comment.

At the moment it says `registeredNodes == desiredMachines && readyAndSchedulableNodes == desiredMachines`.

As a consequence, if `readyAndSchedulableNodes == desiredMachines` it automatically means `registeredNodes == desiredMachines`, because `readyAndSchedulableNodes` is a subset of `registeredNodes`.

Am I missing something? 🤔

Reply:
Basically in case of a rolling update when there is a user pod with a PDB that prevents the node drain. For example:
```
❯ k get machinedeployments.machine.sapcloud.io -A
NAMESPACE             NAME                        READY   DESIRED   UP-TO-DATE   AVAILABLE   AGE
shoot--local--local   shoot--local--local-local   1       1         1            1           28m

❯ k get nodes
NAME                                            STATUS                     ROLES    AGE   VERSION
machine-shoot--local--local-local-57666-9rhtk   Ready                      <none>   10m   v1.32.0
machine-shoot--local--local-local-6cffc-48svp   Ready,SchedulingDisabled   <none>   23m   v1.31.1
```

Reply:
I see, thanks for posting this case. Then it's probably better to keep the check as it is.

Reply:
Maybe it will be good to add something like this above the if:
```
// skip checks if all shoot nodes are ready and match the target node number and if there are no ongoing node drains
```
It took me quite some time to come up with the above case justifying the old if logic
wdyt?